### ISSUE TITLE
How to see the number of iterations and loss changes when running model in mmlspark lightgbm?

### ISSUE THREAD

**Issue by RainFung** [Azure/mmlspark#1094]
Is there UI or log  which to see the number of iterations and loss changes when running model in mmlspark lightgbm?

**Comment by RainFung** [comment]
![image](https://user-images.githubusercontent.com/23121539/122734116-61b10400-d2b0-11eb-89a3-133f664a511d.png)
1 rounds use 1 minite?

**Comment by imatiach-msft** [comment]
@RainFung what is your dataset size and cluster configuration? 1 minute per iteration does seem like a like.  What are the parameters to lightgbm?

**Comment by RainFung** [comment]
I increased the memory of excutor and solved the problem of slow iteration.How can I detect the number of iterations and loss changes like in the local mode？

**Comment by imatiach-msft** [comment]
@RainFung are you using pyspark or scala?  In scala there are a ton of delegates you can override, you can even dynamically change the learning rate while training:

https://github.com/Azure/mmlspark/blob/master/lightgbm/src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMDelegate.scala

in pyspark that's a story that I need to implement.

"How can I detect the number of iterations and loss changes like in the local mode?"

Which functionality is this specifically, can you give more details?  Could you point me to an example in python and I can try to mimic that in pyspark?

**Comment by imatiach-msft** [comment]
this is the delegate parameter:

https://github.com/Azure/mmlspark/blob/master/lightgbm/src/main/scala/com/microsoft/ml/spark/lightgbm/params/LightGBMParams.scala#L463

I think that LightGBMDelegate parameter could be wrapped in pyspark too...

**Comment by RainFung** [comment]
I want to see logs like this in pyspark
```
[LightGBM] [Warning] min_data_in_leaf is set=30, min_child_samples=20 will be ignored. Current value: min_data_in_leaf=30
[LightGBM] [Warning] Accuracy may be bad since you didn't explicitly set num_leaves OR 2^max_depth > num_leaves. (num_leaves=31).
Training until validation scores don't improve for 200 rounds
[500]	train's auc: 0.786176	valid's auc: 0.785457
[1000]	train's auc: 0.78646	valid's auc: 0.785479
Early stopping, best iteration is:
[858]	train's auc: 0.786398	valid's auc: 0.785487
```

**Comment by denmoroz** [comment]
@imatiach-msft 👋 
> in pyspark that's a story that I need to implement.
> I think that LightGBMDelegate parameter could be wrapped in pyspark too...

I think it will be 100% very useful. For instance in my particular scenario: I am using hyperopt to tune booster parameters with `average_precision` as a metric using CV. From the code I see that from Scala I can override [afterTrainIteration](https://github.com/microsoft/SynapseML/blob/339e516e1c38d583c95c95d15d6b09fd86e6a43b/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/LightGBMDelegate.scala#L48) and gather metrics from `validEvalResults` to pass it then to hyperopt fmin / fmax.

But at the same delegate is not available for pyspark users, so what I really have to do is:
1. use LightGBM internal `average_precision` for early stopping
2. then gather predictions for validation set(s) and calculate average precision by my own

It works, but for sure it is not optimal and it would be so much easier (and safer) just to use LightGBM's output.
Having that said I wonder if there are plans to make LightGBMDelegate available for PySpark anytime soon?

Thanks!

**Comment by mtreca** [comment]
I am resurrecting this issue because this is also a pain point of my workflow.
Similarly to @denmoroz, I have to recompute the metric manually after early stopping in order to potentially prune a given trial.
Instead of implementing LightGBMDelegate, which is apparently not in your plans, could we get an attribute getter that would return the evaluation metric from the LGBM model directly?
That would save significant computation time for any form of external parameter tuning.

**Comment by ranadeepsingh** [comment]
Thank you for the clear iteration/loss visibility request. It is related to #2269, but not identical to that concrete logging bug, so this issue remains open as a distinct API requirement. Please keep contributing examples of the desired output.
